concurrent/handler.go

```diff
@@ -0,0 +1,77 @@
+package concurrent
+
+import (
+	"context"
+	"time"
+
+	servicebus "github.com/Azure/azure-service-bus-go"
+	"github.com/Azure/go-shuttle/message"
+	"github.com/Azure/go-shuttle/peeklock"
+	"github.com/devigned/tab"
+)
+
+type Handler struct {
+	messageHandler  message.Handler
+	messages        chan *servicebus.Message
+	lockRenewer     peeklock.LockRenewer
+	renewalInterval *time.Duration
+}
+
+func NewHandler(ctx context.Context, r peeklock.LockRenewer, renewalInterval *time.Duration, handler message.Handler) servicebus.Handler {
+	h := &Handler{
+		renewalInterval: renewalInterval,
+		lockRenewer:     r,
+		messageHandler:  handler,
+		messages:        make(chan *servicebus.Message),
+	}
+	go h.handleMessages(ctx)
+	return h
+}
+
+func (h *Handler) Handle(_ context.Context, msg *servicebus.Message) error {
+	h.messages <- msg
+	// servicebus handler errors interrupt the listener. A single message handling error should not stop the listener.
+	// always return nil here
+	return nil
+}
+
+func (h *Handler) handleMessages(ctx context.Context) {
+	for msg := range h.messages {
+		go h.handleMessage(messageContext(ctx, msg), msg)
```

Comment:
handle each message in a background goroutine

Reply:
I wanted to double check that is that pointless to set a client concurrency limit as it will be always limited by the link credit? I'm assuming we will only worry about the link credit?

Reply:
You could say that AMQP is a push protocol (with flow control) so, as long as you have credits, the broker is pushing message on the TCP connection. They then sit on our side, "unsettled" and we still need to control the concurrency of our workers to process these before their peek lock expires. That is why the link credit (prefetch setting) needs to be balanced with the worker concurrency level to optimize throughput.